clinica/utils/filemanip.py

```diff
@@ -1,209 +1,476 @@
+        }
+    )
     try:
-        data = pandas.DataFrame(
-            {
-                "participant_id": participant_ids,
-                "session_id": session_ids,
-            }
-        )
         data.to_csv(tsv_file, sep="\t", index=False, encoding="utf-8")
     except Exception as e:
         cprint(msg=f"Impossible to save {out_file} with pandas", lvl="error")
         raise e
 
 
+def _raise_non_bids_or_caps_compliant_filename(filename: str) -> None:
+    raise ValueError(
+        f"Input filename {filename} is not in a BIDS or CAPS compliant format."
+        " It does not contain the subject and session information."
+    )
+
+
 def get_subject_id(bids_or_caps_file: str) -> str:
-    """Extract "sub-<participant_id>_ses-<session_label>" from BIDS or CAPS file."""
+    """Extract the subject ID from a BIDS or CAPS file path.
+
+    The subject ID is defined as
+
+        sub-<participant_id>_ses-<session_label>
+
+    In other words, it is the concatenation of the subject and session labels,
+    separated by "_".
+
+    Parameters
+    ----------
+    bids_or_caps_file: str
+        Path to a file from a BIDS or CAPS folder.
+
+    Returns
+    -------
+    subject_id: str
+        The subject ID corresponding to the given file.
+
+    Examples
+    --------
+    >>> get_subject_id("sub-01/ses-M000/pet/sub-01_ses-M000_trc-18FAV45_pet.nii.gz")
+    'sub-01_ses-M000'
+    >>> get_subject_id("sub-01_ses-M000_trc-18FAV45_pet.nii.gz")
+    Traceback (most recent call last):
+    ValueError: Input filename sub-01_ses-M000_trc-18FAV45_pet.nii.gz is not in a BIDS or CAPS compliant format. It does not contain the subject and session information.  # noqa
+
+    See also
+    --------
+    extract_image_ids
+    """
     import re
 
     m = re.search(r"(sub-[a-zA-Z0-9]+)/(ses-[a-zA-Z0-9]+)", bids_or_caps_file)
-
     if not m:
-        raise ValueError(
-            f"Input filename {bids_or_caps_file} is not in a BIDS or CAPS compliant format."
-            " It does not contain the subject and session information."
-        )
-
+        _raise_non_bids_or_caps_compliant_filename(bids_or_caps_file)
     subject_id = m.group(1) + "_" + m.group(2)
 
     return subject_id
 
 
-def get_filename_no_ext(filename):
-    """Get filename without extension [".nii.gz", ".tar.gz", ".niml.dset"]."""
+def get_filename_no_ext(filename: str) -> str:
+    """Get the filename without the extension.
+
+    Parameters
+    ----------
+    filename: str
+        The full filename from which to extract the extension out.
+
+    Returns
+    -------
+    filename_no_ext: str
+        The filename with extension removed.
+
+    Examples
+    --------
+    >>> get_filename_no_ext("foo.nii.gz")
+    'foo'
+    >>> get_filename_no_ext("sub-01/ses-M000/sub-01_ses-M000.tar.gz")
+    'sub-01_ses-M000'
+    """
     from nipype.utils.filemanip import split_filename
 
     _, filename_no_ext, _ = split_filename(filename)
 
     return filename_no_ext
```

Comment:
We can get rid of `nipype` for this now that we have got `pathlib`.

```suggestion
    from pathlib import PurePath
    
    stem = PurePath(filename).stem
    while('.' in stem):
        stem = PurePath(stem).stem

    return stem
```